changelog-generator/src/parse.rs

```diff
@@ -0,0 +1,325 @@
+
+    let branch_call_output = branch_call
+        .output()
+        .expect("Failed git branch --show-current");
+    let branch_name = from_utf8(&branch_call_output.stdout)
+        .expect("Failed to read branch name")
+        .to_string();
+
+    let version_loc = pattern.find(&branch_name).expect(
+        "Could not locate version, check your branch name if it fits the release convention",
+    );
+
+    branch_name[version_loc.range()].to_string()
+}
+
+// 'config' Config
+// 'release_range' a range of 2 versions ex. v3.2.0 , v3.2.1
+pub fn parse_git_log(config: &Config, release_range: (&str, &str)) -> Vec<String> {
+    //git log
+    let mut git_log = process::Command::new("git");
+    // check if
+    if let Some(r_path) = &config.repo_path {
+        git_log.arg("-C").arg(r_path);
+    };
+    git_log.arg("log");
+
+    //add in log range if previous release was found
+    if !release_range.0.trim().is_empty() {
+        git_log.arg(format!("{}..{}", release_range.0, release_range.1));
+    }
+
+    git_log
+        .arg(format!("{}..", release_range.0))
+        .arg("--oneline");
+
+    let git_log_output = git_log.output().expect("Failed git log call");
+    let git_log_str = from_utf8(&git_log_output.stdout).unwrap();
+
+    assert!(!git_log_str.is_empty(), "Git log empty! Make sure the script is ran from the base repo directory or check repository path arg correctness");
+
+    let spl = git_log_str.split("\n");
+    let commit_data: Vec<String> = spl.map(|s| s.into()).collect();
+
+    commit_data
+}
+
+pub fn make_changelog_path(config: &Config) -> String {
+    let changelog_path = match &config.repo_path {
+        Some(path) => format!("{}/{}", path, "CHANGELOG.md"),
+        None => "CHANGELOG.md".to_string(),
+    };
+
+    assert!(
+        Path::new(&changelog_path).exists(),
+        "Could not locate CHANGELOG.md! \
+                                                Make sure script is ran from repo directory \
+                                                or check repository path correctness"
+    );
+    changelog_path
+}
+
+// parse the git log and collect the commit data based on it with github API calls
+// `input` is a vector of every line from gitlog
+// 'login_info' is the login info of the caller (username:pass/authtoken) needed to make
+// calls to the API without getting timed out or limited in the number of calls per hour
+pub fn parse_commits(input: Vec<String>, login_info: (&str, &str)) -> Vec<Commit> {
+    let mut commits: Vec<Commit> = vec![];
+    let pr_id_pattern = regex::Regex::new(r"(#[0-9]+)").expect("Invalid regex");
+    for commit_str in input.iter() {
+        let mut splitter = commit_str.split_whitespace();
+
+        // need it for its length
+        let commit_id = splitter.next().unwrap();
+
+        let pr_id_str = splitter.last().unwrap();
+
+        // skip pull requests that have no Pull Request ID
+        if !pr_id_pattern.is_match(pr_id_str) {
+            continue;
```

Comment:
tbh. this should be a `panic!`

We do not allow merges to manta without a corresponding PR and having a merge like that should be considered broken state that needs to be reverted.

This constraint also covers your edit in 
https://github.com/Manta-Network/dev-tools/pull/5/files#diff-f3388a952b5050ccfd53ffc53f641866dd31eb91fd3b3b008a79f63a71dc1d08R155

Reply:
Ok right now things are quite chaotic, let's say it happens again and we are in the middle of a freeze, how would we handle that kind of merge, which would be making the changelog generator panic?

Reply:
As a compromise, you could have it return a non-zero exit code in case it encounters this.

Then the github workflow can decide how it wants to handle the exit code - fail on anything non-zero or ignore specifc cases
Also, we can always create the changelog manually in case this fails.
This is a 80/20 problem, it's simply not worth covering every exception.